server/src/main/java/org/opensearch/cluster/metadata/Metadata.java

```diff
@@ -1425,6 +1446,45 @@ public Builder generateClusterUuidIfNeeded() {
         }
 
         public Metadata build() {
+            TimeValue buildStartTime = TimeValue.timeValueMillis(System.nanoTime());
+            boolean recomputeRequired = indices.equals(indicesPreviousState) == false || customs.equals(customsPreviousState) == false;
+            TimeValue recomputeEndTime = TimeValue.timeValueMillis(System.nanoTime());
+            logger.info(
+                "Recompute required: {}, time taken for comparing indices: {} ms",
+                recomputeRequired,
+                (recomputeEndTime.getNanos() - buildStartTime.getNanos()) / 1000000L
+            );
+            // Will simplify this later to omit this recomputeRequired variable entirely - it's only used for testing for now
+            if (recomputeRequired) {
+                buildMetadataIndicesLookups();
+            }
+
+            Metadata metadata = new Metadata(
+                clusterUUID,
+                clusterUUIDCommitted,
+                version,
+                coordinationMetadata,
+                transientSettings,
+                persistentSettings,
+                hashesOfConsistentSettings,
+                indices,
+                templates,
+                customs,
+                allIndices,
+                visibleIndices,
+                allOpenIndices,
+                visibleOpenIndices,
+                allClosedIndices,
+                visibleClosedIndices,
+                indicesLookup
+            );
+            TimeValue endBuildTime = TimeValue.timeValueMillis(System.nanoTime());
+            // Logging for testing only - will remove in future iterations
+            logger.info("built metadata in {} ms", (endBuildTime.millis() - buildStartTime.millis()) / 1000000L);
+            return metadata;
+        }
+
+        private void buildMetadataIndicesLookups() {
             // TODO: We should move these datastructures to IndexNameExpressionResolver, this will give the following benefits:
```

Comment:
Did we explore about this ToDo ?? 

Is it possible to move anything as expression resolver ?

Reply:
We need a source of truth of `indices` and `customs` to evaluate the other data-structures which is presently in `ClusterState`.  We will light-weight ClusterState to some extent but will add an overhead in index, etc name resolution methods. 

Also, parallel developments on this: https://github.com/opensearch-project/OpenSearch/issues/7002#issuecomment-1573277658 but this will be outside the scope of this PR/issue. 

I intend to keep the scope of this fix to avoid unnecessary computing all these data-structures again when they are not required to reduce the ClusterState computation time without adding addtional memory footprint.